flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -0,0 +1,186 @@
+			maxCheckpoints
+		);
+	}
+
+	ChannelStateWriterImpl(ConcurrentMap<Long, ChannelStateWriteResult> results, ChannelStateWriteRequestExecutor executor, int maxCheckpoints) {
+		this.results = results;
+		this.maxCheckpoints = maxCheckpoints;
+		this.executor = executor;
+	}
+
+	@Override
+	public void start(long checkpointId, CheckpointOptions checkpointOptions) {
+		LOG.debug("start checkpoint {} ({})", checkpointId, checkpointOptions);
+		ChannelStateWriteResult result = new ChannelStateWriteResult();
+		ChannelStateWriteResult put = results.computeIfAbsent(checkpointId, id -> {
+			Preconditions.checkState(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);
+			enqueue(new CheckpointStartRequest(checkpointId, result, checkpointOptions.getTargetLocation()), false);
+			return result;
+		});
+		Preconditions.checkArgument(put == result, "result future already present for checkpoint id: " + checkpointId);
+	}
+
+	@Override
+	public void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add input data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)), false);
+	}
+
+	@Override
+	public void addOutputData(long checkpointId, ResultSubpartitionInfo info, int startSeqNum, Buffer... data) {
+		LOG.debug("add output data, checkpoint id: {}, channel: {}, startSeqNum: {}, num buffers: {}",
+			checkpointId, info, startSeqNum, data == null ? 0 : data.length);
+		enqueue(write(checkpointId, info, checkBufferType(data)), false);
+	}
+
+	@Override
+	public void finishInput(long checkpointId) {
+		LOG.debug("finish input data, checkpoint id: {}", checkpointId);
+		enqueue(completeInput(checkpointId), false);
+	}
+
+	@Override
+	public void finishOutput(long checkpointId) {
+		LOG.debug("finish output data, checkpoint id: {}", checkpointId);
+		enqueue(completeOutput(checkpointId), false);
+	}
+
+	@Override
+	public void abort(long checkpointId, Throwable cause) {
+		LOG.debug("abort, checkpoint id: {}", checkpointId);
+		enqueue(ChannelStateWriteRequest.abort(checkpointId, cause), true); // abort already started
+		enqueue(ChannelStateWriteRequest.abort(checkpointId, cause), false); // abort enqueued but not started
+	}
+
+	@Override
+	public ChannelStateWriteResult getWriteResult(long checkpointId) {
+		LOG.debug("requested write result, checkpoint id: {}", checkpointId);
+		ChannelStateWriteResult result = results.remove(checkpointId);
+		Preconditions.checkArgument(result != null, "channel state write result not found for checkpoint id " + checkpointId);
+		return result;
+	}
+
+	public void open() {
+		executor.start();
+	}
+
+	@Override
+	public void close() throws IOException {
```

Comment:
This can be done only via `SubtaskCheckpointCoordinatorImpl#close`, but i do not see `StreamTask` or other places will call `SubtaskCheckpointCoordinatorImpl#close` from the current codes. What is the consideration for this?

Reply:
`SubtaskCheckpointCoordinatorImpl` is registered with `CloseableRegistry` in its constructor (which in turn is closed by `StreamTask`).